File: tools/test_menu_load.lua
-- Smoke test for Features/Menu.lua M.load() merging persisted settings
local function ensure_dir(dir)
    if package.config:sub(1,1) == "\\" then
        os.execute(('if not exist "%s" mkdir "%s"'):format(dir, dir))
    else
        os.execute(('mkdir -p "%s"'):format(dir))
    end
end

local dir = 'game data'
ensure_dir(dir)
local path = dir .. '/settings.json'

local content = [[
{
  AutoUpgrade = false,
  Settings = {
    Dungeon = {
      Enabled = true,
      AutoSelectHighest = true
    }
  }
}
]]

local fh, err = io.open(path, 'wb')
if not fh then
    print('Cannot write settings file:', err)
    os.exit(2)
end
fh:write(content)
fh:close()

-- Load the module (it will attempt to pcall(M.load) at the end)
local ok, M = pcall(function() return dofile('Features/Menu.lua') end)
if not ok or type(M) ~= 'table' then
    print('Failed to load Menu module:', M)
    os.exit(2)
end

local passed = true
local function expect(path, expected)
    local got = M.get(path)
    if got ~= expected then
        print('FAIL', path, 'expected', tostring(expected), 'got', tostring(got))
        passed = false
    else
        print('OK', path, tostring(got))
    end
end

expect('AutoUpgrade', false)
expect('Settings.Dungeon.Enabled', true)
expect('Settings.Dungeon.AutoSelectHighest', true)

if passed then
    print('SMOKE TEST PASSED')
    os.exit(0)
else
    print('SMOKE TEST FAILED')
    os.exit(1)
end

File: World Zero/tools/test_menu_load.py
import os, json, sys

DATA_DIR = os.path.join(os.getcwd(), 'game data')
os.makedirs(DATA_DIR, exist_ok=True)
path = os.path.join(DATA_DIR, 'settings.json')

# persisted settings we want to test
persisted = {
    "AutoUpgrade": False,
    "Settings": {
        "Dungeon": {
            "Enabled": True,
            "AutoSelectHighest": True
        }
    }
}
with open(path, 'w', encoding='utf-8') as fh:
    json.dump(persisted, fh, indent=2)
print('Wrote', path)

# defaults as in Features/Menu.lua (subset for the test)
def defaults():
    return {
        'AutoUpgrade': True,
        'RemoveDamage': False,
        'AutoChest': True,
        'CoinMagnet': True,
        'Settings': {
            'Dungeon': {
                'Enabled': False,
                'AutoSelectHighest': False,
                'CustomDungeon': { 'DungeonId': 27 }
            },
            'Tower': {
                'Enabled': False,
                'AutoSelectHighest': False,
                'CustomTower': { 'TowerId': 27 }
            },
            'AutoRejoin': { 'Enabled': True, 'Delay': 5 },
            'AutoSell': { 'Enabled': False, 'Common': False, 'Uncommon': False, 'Rare': False, 'Epic': False }
        }
    }

# merge_defaults behavior from Lua: only set values from src into target when target[k] == nil
# We'll simulate that in Python

def merge_defaults(target, src):
    for k, v in src.items():
        if isinstance(v, dict):
            if not isinstance(target.get(k), dict):
                target[k] = {}
            merge_defaults(target[k], v)
        else:
            if target.get(k) is None:
                target[k] = v

# Simulate initial config: deep copy of empty env then merge defaults
config = {}
merge_defaults(config, defaults())
print('\nConfig after applying defaults (before load):')
print(json.dumps(config, indent=2))

# Now simulate loading persisted file and calling merge_defaults(config, persisted)
with open(path, 'r', encoding='utf-8') as fh:
    data = json.load(fh)

merge_defaults(config, data)
print('\nConfig after merge_defaults(config, persisted):')
print(json.dumps(config, indent=2))

# Report expected vs actual for the test assertions (mimic original Lua test expectations)
checks = [
    ('AutoUpgrade', False),
    ('Settings.Dungeon.Enabled', True),
    ('Settings.Dungeon.AutoSelectHighest', True)
]

print('\nAssertions (expected -> actual):')
all_ok = True
for path, expected in checks:
    cur = config
    for p in path.split('.'):
        cur = cur.get(p) if isinstance(cur, dict) else None
    print(path, 'expected=', expected, 'actual=', cur)
    if cur != expected:
        all_ok = False

if all_ok:
    print('\nSMOKE TEST PASSED: persisted values were applied')
    sys.exit(0)
else:
    print('\nSMOKE TEST SHOWS persisted values did NOT OVERRIDE defaults when target already had values (this matches merge_defaults behavior).')
    sys.exit(2)
